### PR TITLE
Restrict access to create session

### PR DESCRIPTION
**Pull request by swemail**
With this the backend won't allow creating sessions without being a `publicHost`

### CODE REVIEW

**Comment by swemail** [comment]
> Wouldn't we use that for creating private sessions? which everyone will be able to do? 🤔

You're right, it was not there when I started this 😄 Let's reiterate this. Maybe we need different enpoints?

File: functions/src/api/lib/restrictAccessToRole.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import {Next} from 'koa';
+import {getAuth} from 'firebase-admin/auth';
+import {ROLES} from '../../../../shared/src/types/User';
+import {FirebaseAuthContext} from './firebaseAuth';
+
+const restrictAccessToRole =
+  (role: keyof typeof ROLES) =>
+  async (ctx: FirebaseAuthContext, next: Next) => {
+    const customClaims = (await getAuth().getUser(ctx.user.id)).customClaims;
```
Comment (gewfy) [inline functions/src/api/lib/restrictAccessToRole.ts:9]: Can't we add the claims on the user context with the existing middleware instead? And keep this one to just check that object?

File: functions/src/api/sessions/index.ts [hunk 2]
```diff
@@ -36,18 +37,23 @@ const CreateSessionSchema = yup.object().shape({
...
-    contentId,
-    type,
-    startTime,
-    language,
-  });
-});
+sessionsRouter.post(
+  '/',
+  validator({body: CreateSessionSchema}),
+  restrictAccessToRole('publicHost'),
```
Comment (gewfy) [inline functions/src/api/sessions/index.ts:43]: Really like this way of validating 🙌

**Review by gewfy — CHANGES_REQUESTED** [review]

File: functions/src/api/lib/restrictAccessToRole.ts (hunk 1 above)
Comment (swemail) [reply functions/src/api/lib/restrictAccessToRole.ts:9]: Super improvement, thanks!

**Comment by Nipher** [comment]
or maybe it is possible to check the session type that was sent through the endpoint and check the necessary user role accordingly?

Either that or two endpoints sounds good to me, no strong opinions

**Review by Nipher — APPROVED** [review]

**Review by gewfy — APPROVED** [review]
Really nice! 🙌

**Pull request merged**